client/src/components/LevelLayout.tsx

```diff
@@ -46,6 +57,19 @@ const LevelLayout: React.FC<Props> = ({
     didWin();
   }
 
+  const [algo, setAlgo] = useState<number>(0);
+  const [level, setLevel] = useState<number>(0);
+  const [customRestart, setCustomRestart] = useState<boolean>(false);
+  const navigate = useNavigate();
+
+  function generateLevels() {
```

Comment:
For performance, it's not ideal to have function calls return DOM elements as the function will be called on each render even in the return of the function stays the same. Generally speaking, this is a bad practice, see [here](https://kentcdodds.com/blog/dont-call-a-react-function-component).

Instead, consider using `useMemo`. 

This is super, duper not necessary though. I wouldn't be peeved if you merged as is.

Reply:
updated